pkg/volume/csi/csi_attacher.go

```diff
@@ -269,6 +269,25 @@ func (c *csiAttacher) MountDevice(spec *volume.Spec, devicePath string, deviceMo
 		return err
 	}
 
+	dataDir := filepath.Dir(deviceMountPath)
+	if err := os.MkdirAll(dataDir, 0750); err != nil {
+		glog.Error(log("attacher.MountDevice failed to create dir %#v:  %v", dataDir, err))
+		return err
+	}
+
+	data := map[string]string{
+		volDataKey.volHandle:  csiSource.VolumeHandle,
+		volDataKey.driverName: csiSource.Driver,
+	}
+	if err := saveVolumeData(dataDir, volDataFileName, data); err != nil {
+		glog.Error(log("failed to save volume info data: %v", err))
+		if err := os.RemoveAll(dataDir); err != nil {
+			glog.Error(log("failed to remove dir after error [%s]: %v", dataDir, err))
+			return err
+		}
+		return err
+	}
+
```

Comment:
If the `MountDevice(...)` operation fails below, how is the save file cleaned up?

Reply:
If `MounDevice` fails, then the whole directory is cleaned by `removeMountDir`. But you raised valid point, there are other error paths (e.g. failed to load secrets) that lead to uncleaned json file. I'll fix it.

Reply:
Filled #65323 + PR